src/context/TodosContext.tsx

```diff
@@ -0,0 +1,220 @@
+      return { ...state, isLoading: false, todos: action.payload };
+
+    case 'todos/setFilterStatus':
+      return {
+        ...state,
+        filterStatus: action.payload,
+      };
+
+    case 'todos/setTempTodo':
+      return {
+        ...state,
+        tempTodo: action.payload,
+      };
+    case 'todos/toggleAll':
+      return {
+        ...state,
+        todos: state.todos.map(todo => ({
+          ...todo,
+          completed: !state.todos.every(todoItem => todoItem.completed),
+        })),
+      };
+
+    case 'todos/setError':
+      return { ...state, todosError: action.payload };
+
+    case 'todos/removeError':
+      return { ...state, todosError: '' };
+
+    case 'rejected':
+      return { ...state, isLoading: false, todosError: action.payload };
+    default:
+      return state;
+  }
+}
+
+const TodosProvider: React.FC<Props> = ({ children }) => {
+  const [state, dispatch] = useReducer(reducer, initialState);
+  const {
+    todos,
+    filterStatus,
+    isDeletingAllCompleted,
+    isLoading,
+    todosError,
+    tempTodo,
+  } = state;
+
+  const handleRemoveError = () => {
+    dispatch({ type: 'todos/removeError' });
+  };
+
+  useEffect(() => {
+    const fetchTodos = async () => {
+      dispatch({ type: 'loading', payload: true });
+      try {
+        const fetchedTodos = await getTodos();
+
+        dispatch({ type: 'todos/loaded', payload: fetchedTodos });
+      } catch {
+        dispatch({
+          type: 'rejected',
+          payload: 'Unable to load todos',
+        });
+
+        wait(3000).then(handleRemoveError);
+      } finally {
+        dispatch({ type: 'loading', payload: false });
+      }
+    };
+
+    fetchTodos();
+  }, []);
+
+  const addTodo = (todo: Todo) => {
+    dispatch({ type: 'todos/addTodo', payload: todo });
+  };
+
+  const handleDeleteTodo = (id: number) => {
+    dispatch({ type: 'todos/delete', payload: id });
+  };
+
+  const setTempTodo = (todo: Todo | null) => {
+    dispatch({ type: 'todos/setTempTodo', payload: todo });
+  };
+
+  const handleFilterTodo = (status: FilterStatus) => {
+    dispatch({ type: 'todos/setFilterStatus', payload: status });
+  };
+
+  const toggleAllTodo = () => {
+    dispatch({ type: 'todos/toggleAll' });
+  };
+
+  const handleSetError = (errorMessage: string) => {
+    dispatch({ type: 'todos/setError', payload: errorMessage });
+
+    wait(3000).then(handleRemoveError);
+  };
+
+  return (
+    <TodosContext.Provider
+      value={{
+        addTodo,
+        todos,
+        tempTodo,
+        isLoading,
+        todosError,
+        filterStatus,
+        isDeletingAllCompleted,
+        setTempTodo,
+        handleDeleteTodo,
+        handleRemoveError,
+        handleFilterTodo,
+        toggleAllTodo,
+        handleSetError,
+        dispatch,
+      }}
+    >
+      {children}
+    </TodosContext.Provider>
+  );
+};
+
+const useTodos = () => {
+  const context = useContext(TodosContext);
+
+  if (context === undefined) {
```

Comment:
```suggestion
  if (!context) {
```
You don't need to compare with `undefined`
https://developer.mozilla.org/en-US/docs/Glossary/Falsy